30413/Mihai Alexandru Sandu/Lab03/extra xor linked list/main.c

```diff
@@ -0,0 +1,305 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <inttypes.h>
+
+
+/* to get the address o the current node:   address of previous node XOR address of the next node
+   pt parcurgere ->  address of current ( care este address of previous node XOR address of the next node) XOR address of previous node
+*/
+
+typedef struct node
+    {
+        int value;
+        struct node *npx;
+    }NodeT;
+
+typedef struct SLL
+{
+    int count;
+    NodeT *first;
+    NodeT *last;
+}SLListT;
+
+SLListT *createEmptySLList()
+{
+    SLListT *list = (SLListT*)malloc(sizeof(SLListT));
+    if (list)
+    {
+        list->count = 0;
+        list->first = list->last = NULL;
+    }
+    return list;
+}
+
+NodeT *createNode(int value)
+{
+    NodeT *p = (NodeT*)malloc(sizeof(NodeT));
+    if (p)
+    {
+        p->npx = NULL;
+        p->value = value;
+    }
+    return p;
+}
+NodeT *XOR(NodeT *a,NodeT *b)
+{
+    return (NodeT*)((uintptr_t)(a) ^ (uintptr_t)(b));
+}
+
+void AddFirst(SLListT *list, NodeT *p)
+{
+    if (list)
+    {
+        if (list->count == 0)
+        {
+            list->first = list->last = p;
+        }
+        else if (list->count == 1)
+        {
+            p->npx = XOR(NULL,list->first);
+            list->first->npx = XOR(p,NULL);
+            list->first = p;
+        }
+        else
+        {
+            p->npx = XOR(NULL,list->first);
+            list->first->npx = XOR(p,list->first->npx);
+            list->first = p;
+        }
+        list->count++;
+    }
+}
+
+void AddLast(SLListT *list, NodeT *p)
+{
+    if(list)
+    {
+        if (list->count == 0)
+        {
+            list->first = list->last = p;
+        }
+        else if (list->count == 1)
+        {
+            p->npx = XOR(list->last,NULL);
+            list->first->npx =XOR(NULL,p);
+            list->last = p;
+        }
+        else
+        {
+            p->npx = XOR(list->last,NULL);
+            list->last->npx = XOR(list->last,p);
+            list->last = p;
+        }
+        list->count++;
+    }
+}
+
+void InsertBeforeKey(int key, SLListT *list, NodeT *node)
```

Comment:
This functions works but it doesn't work. When I print the list after I inserted an element, it prints the list corectly but it gives me an error. I think somethink about memory

Reply:
I can't help you with debugging the code, sorry!